### PR TITLE
refactor(grid): make screen rendering more multibyte than ever before

### PR DESCRIPTION
**Pull request by bfredl**
Problem: buffer text with composing chars are converted from UTF-8 to an array of up to seven UTF-32 values and then converted back to UTF-8 strings.

Solution: Convert buffer text directly to UTF-8 based `schar_T` values.

The limit of the text size is now in `schar_T` bytes, which is currently 31+1 but easily could be raised as it no longer multiplies the size of the entire screen grid when not used, the full size is only required for temporary scratch buffers.

Also does some general cleanup to `win_line` text handling, which was unnecessarily complicated due to multi-byte rendering being an "opt-in" feature long ago. Nowadays, a char is just a char, regardless if it consists of one ASCII byte or multiple bytes.

### CODE REVIEW

File: src/nvim/grid.c [hunk 1]
```diff
@@ -254,11 +239,10 @@ void line_do_arabic_shape(schar_T *buf, int cols)
...
     if (rest + (size_t)off + 1 > MAX_SCHAR_SIZE) {
-      // TODO(bfredl): this cannot happen just yet, as we only construct
-      // schar_T values with up to MAX_MCO+1 composing codepoints. When code
-      // is improved so that MAX_SCHAR_SIZE becomes the only/sharp limit,
-      // we need be able to peel off a composing char which doesn't fit anymore.
-      abort();
+      // Too bigly, discard one code-point.
+      // This should be enough as c0 cannot grow more than from 2 to 4 bytes
+      // (base arabic to extended arabic)
+      rest -= (size_t)utf_cp_head_off(scbuf + off, scbuf + off + rest - 1) + 1;
```
Comment (zeertzjq) [inline src/nvim/grid.c:245]: This looks really strange. I think two overflows happen on this line, one in the `(size_t)` cast and one in the subtraction.

Comment (bfredl) [reply src/nvim/grid.c:245]: right, the function tries to be "helpful" and returns an already negated value.

File: test/functional/ui/fold_spec.lua [hunk 2]
```diff
@@ -1156,7 +1154,7 @@ describe("folded lines", function()
           [2:---------------------------------------------]|
           [3:---------------------------------------------]|
         ## grid 2
-          {5:^+--  2 lines: å 语 x̎͂̀̂͛͛ ﺎﻠﻋَﺮَﺒِﻳَّﺓ·················}|
+          {5:^+--  2 lines: å 语 x̨̣̘̫̲͚͎̎͂̀̂͛͛̾͢ ﺎﻠﻋَﺮَﺒِﻳَّﺓ·················}|
```
[the remark this answers is not in the document]
Comment (bfredl) [reply test/functional/ui/fold_spec.lua:1157]: did so

File: runtime/doc/news.txt [hunk 3]
```diff
@@ -290,6 +290,13 @@ The following changes to existing APIs or features add new behavior.
   Note that syntax highlighting of code examples requires a matching parser
   and may be affected by custom queries.
 
+• Support for rendering mulibyte characters using composing characters has been
+  enhanced. The maximum limit have beeen increased from 1+6 codepoints to
+  31 bytes, which is guaranteed to fit all chars from before but often more.
+
+  NOTE: the rexexp engine still has a hard-coded limit of considering
```
Comment (zeertzjq) [inline runtime/doc/news.txt:297]:
```suggestion
  NOTE: the regexp engine still has a hard-coded limit of considering
```

File: runtime/doc/news.txt [hunk 4]
```diff
@@ -290,6 +290,13 @@ The following changes to existing APIs or features add new behavior.
   Note that syntax highlighting of code examples requires a matching parser
   and may be affected by custom queries.
 
+• Support for rendering mulibyte characters using composing characters has been
```
Comment (clason) [inline runtime/doc/news.txt:293]:
```suggestion
• Support for rendering multibyte characters using composing characters has been
```

File: runtime/doc/news.txt [hunk 5]
```diff
@@ -290,6 +290,13 @@ The following changes to existing APIs or features add new behavior.
...
+  enhanced. The maximum limit have beeen increased from 1+6 codepoints to
```
Comment (clason) [inline runtime/doc/news.txt:294]:
```suggestion
  enhanced. The maximum limit have been increased from 1+6 codepoints to
```

File: runtime/doc/mbyte.txt [hunk 6]
```diff
@@ -646,7 +646,8 @@ widespread as file format.
 A composing or combining character is used to change the meaning of the
 character before it.  The combining characters are drawn on top of the
 preceding character.
-Up to six combining characters can be displayed.
+Too big combined characters cannot be displayed, but they can still be
+introspected using the |g8| and |ga| commands desribed below.
```
Comment (zeertzjq) [inline runtime/doc/mbyte.txt:650]:
```suggestion
inspected using the |g8| and |ga| commands desribed below.
```

File: runtime/doc/vim_diff.txt [hunk 7]
```diff
@@ -722,9 +722,16 @@ Options:
...
   *'maxcombine'* *'mco'*
-    Nvim always displays up to 6 combining characters.  You can still edit
-    text with more than 6 combining characters, you just can't see them.
-    Use |g8| or |ga|.  See |mbyte-combining|.
+    Nvim counts maximum character sizes in bytes, not codepoints. This is
+    guaranteed to be big enough to always fit all chars properly displayed
+    in vim with 'maxcombine' set to 6.
+
+    You can still edit text with larger characters than fits in the screen buffer,
+    you just can't see them. Use |g8| or |ga|.  See |mbyte-combining|.
```
Comment (zeertzjq) [inline runtime/doc/vim_diff.txt:730]:
```suggestion
    you just can't see them. Use |g8| or |ga|. See |mbyte-combining|.
```

File: src/nvim/message.c [hunk 8]
```diff
@@ -1560,12 +1560,6 @@ int msg_outtrans_len(const char *msgstr, int len, int attr)
     mode_displayed = false;
   }
 
-  // If the string starts with a composing character first draw a space on
-  // which the composing char can be drawn.
-  if (utf_iscomposing(utf_ptr2char(msgstr))) {
-    msg_puts_attr(" ", attr);
-  }
-
```
Comment (zeertzjq) [inline src/nvim/message.c:1568]: This change seems wrong. The code below doesn't seem to involve `schar_from_buf_first()`, so this is not done in the code below.

Comment (bfredl) [reply src/nvim/message.c:1568]: `msg_puts_len` -> `msg_puts_display` -> `grid_line_puts` -> `utfc_ptr2schar_len` -> `schar_from_buf_first`

All multibyte rendering ends up in `utfc_ptr2schar[_len]` eventually, that's the point of these refactors..

Comment (zeertzjq) [reply src/nvim/message.c:1568]: Ah yes, my mistake.

**Pull request merged**